File: spacevil_21052019/java_src/Shader.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Flags.ShaderType;
import static org.lwjgl.opengl.GL20.*;

import java.io.FileReader;
import java.io.BufferedReader;

final class Shader {

    private int _program_id;

    long getProgramID() {
        return _program_id;
    }

    private int _shader_vertex;
    private String _code_vertex;

    private int _shader_fragment;
    private String _code_fragment;

    private int _shader_geometry;
    private String _code_geometry;

    private int _shader_compute;
    private String _code_compute;

    String getCode(ShaderType shader_type) {
        switch (shader_type) {
        case VERTEX:
            return _code_vertex;

        case FRAGMENT:
            return _code_fragment;

        case GEOMETRY:
            return _code_geometry;

        case COMPUTE:
            return _code_compute;

        default:
            break;
        }
        return "";
    }

    void setShaderPath(ShaderType shader_type, String path) {
        String code = readSource(path);

        switch (shader_type) {
        case VERTEX:
            _code_vertex = code.toString();
            break;

        case FRAGMENT:
            _code_fragment = code.toString();
            break;

        case GEOMETRY:
            _code_geometry = code.toString();
            break;

        case COMPUTE:
            _code_compute = code.toString();
            break;

        default:
            break;
        }
    }

    Shader() {
        initDefaults();
    }

    Shader(String vertex_code, String fragmend_code) {
        _code_vertex = vertex_code;
        _code_fragment = fragmend_code;

        // defaults
        _code_geometry = "";
        _code_compute = "";
    }

    Shader(BufferedReader vertex_code, BufferedReader fragmend_code) {
        _code_vertex = readSource(vertex_code);
        _code_fragment = readSource(fragmend_code);

        // defaults
        _code_geometry = "";
        _code_compute = "";
    }

    private void initDefaults() {
        _code_vertex = "";
        _code_fragment = "";
        _code_geometry = "";
        _code_compute = "";
    }

    private String readSource(BufferedReader source) {
        StringBuilder code = new StringBuilder();
        String line = null;
        try {
            while ((line = source.readLine()) != null) {
                code.append(line);
                code.append('\n');
            }
        } catch (Exception e) {
            System.out.println("The stream could not be read:");
            System.out.println(e.getMessage());
        }
        return code.toString();
    }

    private String readSource(String source) {
        StringBuilder code = new StringBuilder();
        String line = null;
        try {
            BufferedReader reader = new BufferedReader(new FileReader(source));
            while ((line = reader.readLine()) != null) {
                code.append(line);
                code.append('\n');
            }
            reader.close();
        }

        catch (Exception e) {
            System.out.println("The file could not be read:");
            System.out.println(e.getMessage());
        }
        return code.toString();
    }

    void compile() {
        _program_id = glCreateProgram();
        // compiling
        if (!_code_vertex.equals("")) {
            _shader_vertex = compileShader(ShaderType.VERTEX, _code_vertex);
            glAttachShader(_program_id, _shader_vertex);
        }

        if (!_code_fragment.equals("")) {
            _shader_fragment = compileShader(ShaderType.FRAGMENT, _code_fragment);
            glAttachShader(_program_id, _shader_fragment);
        }

        if (!_code_geometry.equals("")) {
            _shader_geometry = compileShader(ShaderType.GEOMETRY, _code_geometry);
            glAttachShader(_program_id, _shader_geometry);
        }

        if (!_code_compute.equals("")) {
            _shader_compute = compileShader(ShaderType.COMPUTE, _code_compute);
            glAttachShader(_program_id, _shader_compute);
        }
        // linking
        glLinkProgram(_program_id);

    }

    private int compileShader(ShaderType shader_type, String code) {
        int shader_id = glCreateShader(shader_type.getValue());
        glShaderSource(shader_id, code);
        glCompileShader(shader_id);
        return shader_id;
    }

    void useShader() {
        glUseProgram(_program_id);
    }

    void deleteShader() {
        detachShadersCode();
        deleteShadersID();
        glDeleteProgram(_program_id);
    }

    private void detachShadersCode() {
        glDetachShader(_program_id, _shader_vertex);
        glDetachShader(_program_id, _shader_fragment);
        glDetachShader(_program_id, _shader_geometry);
        glDetachShader(_program_id, _shader_compute);
    }

    private void deleteShadersID() {
        glDeleteShader(_shader_vertex);
        glDeleteShader(_shader_fragment);
        glDeleteShader(_shader_geometry);
        glDeleteShader(_shader_compute);
    }
}

File: SpaceVIL/Engine/Shader.cs
// #define LINUX 

using System;
using System.Text;
using System.IO;
using static OpenGL.OpenGLWrapper;

namespace SpaceVIL
{
    internal sealed class Shader// : GL.WGL.OpenWGL
    {
        const uint Vertex = 0x8B31;
        const uint Fragment = 0x8B30;
        const uint Geometry = 0x8DD9;
        const uint Compute = 0x91B9;


        private uint _programId;
        internal uint GetProgramID()
        {
            return _programId;
        }

        private uint _shaderVertex;
        private String _codeVertex;

        private uint _shaderFragment;
        private String _codeFragment;

        private uint _shaderGeometry;
        private String _codeGeometry;

        private uint _shaderCompute;
        private String _codeCompute;

        internal String GetCode(uint shaderType)
        {
            switch (shaderType)
            {
                case Vertex:
                    return _codeVertex;

                case Fragment:
                    return _codeFragment;

                case Geometry:
                    return _codeGeometry;

                case Compute:
                    return _codeCompute;

                default:
                    break;
            }
            return String.Empty;
        }

        internal void SetShaderPath(uint shaderType, String path)
        {
            String code = ReadSource(path);

            switch (shaderType)
            {
                case Vertex:
                    _codeVertex = code.ToString();
                    break;

                case Fragment:
                    _codeFragment = code.ToString();
                    break;

                case Geometry:
                    _codeGeometry = code.ToString();
                    break;

                case Compute:
                    _codeCompute = code.ToString();
                    break;

                default:
                    break;
            }
        }

        private String _name;
        internal String GetShaderName()
        {
            return _name;
        }

        internal Shader(String designation)
        {
            _name = designation;
            InitDefaults();
        }

        internal Shader(String vertexCode, String fragmendCode, String designation)
        {
            _name = designation;

            _codeVertex = vertexCode;
            _codeFragment = fragmendCode;

            //defaults
            _codeGeometry = String.Empty;
            _codeCompute = String.Empty;
        }

        internal Shader(Stream vertexCode, Stream fragmendCode, String designation)
        {
            _name = designation;
            _codeVertex = ReadSource(vertexCode);
            _codeFragment = ReadSource(fragmendCode);

            //defaults
            _codeGeometry = String.Empty;
            _codeCompute = String.Empty;
        }

        internal Shader(Stream vertexCode, Stream geometryCode, Stream fragmendCode, String designation)
        {
            _name = designation;
            _codeVertex = ReadSource(vertexCode);
            _codeGeometry = ReadSource(geometryCode);
            _codeFragment = ReadSource(fragmendCode);

            //defaults
            _codeCompute = String.Empty;
        }

        private void InitDefaults()
        {
            _codeVertex = String.Empty;
            _codeFragment = String.Empty;
            _codeGeometry = String.Empty;
            _codeCompute = String.Empty;
        }

        private String ReadSource(Stream source)
        {
            StringBuilder code = new StringBuilder();
            try
            {
                using (StreamReader sr = new StreamReader(source))
                {
                    while (!sr.EndOfStream)
                        code.Append(sr.ReadLine() + "\n");
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("The stream could not be read:");
                Console.WriteLine(e.Message);
            }
            return code.ToString();
        }

        private String ReadSource(String source)
        {
            StringBuilder code = new StringBuilder();
            try
            {
                using (StreamReader sr = new StreamReader(source))
                {
                    while (!sr.EndOfStream)
                        code.Append(sr.ReadLine() + "\n");
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("The file could not be read:");
                Console.WriteLine(e.Message);
            }
            return code.ToString();
        }

        internal void Compile()
        {
            _programId = glCreateProgram();
            //compiling
            if (!_codeVertex.Equals(String.Empty))
            {
                _shaderVertex = CompileShader(Vertex, _codeVertex);
                glAttachShader(_programId, _shaderVertex);
            }

            if (!_codeFragment.Equals(String.Empty))
            {
                _shaderFragment = CompileShader(Fragment, _codeFragment);
                glAttachShader(_programId, _shaderFragment);
            }

            if (!_codeGeometry.Equals(String.Empty))
            {
                _shaderGeometry = CompileShader(Geometry, _codeGeometry);
                glAttachShader(_programId, _shaderGeometry);
            }

            if (!_codeCompute.Equals(String.Empty))
            {
                _shaderCompute = CompileShader(Compute, _codeCompute);
                glAttachShader(_programId, _shaderCompute);
            }
            //linking
            glLinkProgram(_programId);    
        }

        private uint CompileShader(uint shaderType, String code)
        {
            uint shaderId = glCreateShader((uint)shaderType);
            glShaderSource(shaderId, 1, new[] { code }, new[] { code.Length });
            glCompileShader(shaderId);
            return shaderId;
        }

        internal void UseShader()
        {
            glUseProgram(_programId);
        }

        internal void DeleteShader()
        {
            DetachShadersCode();
            DeleteShadersID();
            glDeleteProgram(_programId);
        }

        private void DetachShadersCode()
        {
            glDetachShader(_programId, _shaderVertex);
            glDetachShader(_programId, _shaderFragment);
            glDetachShader(_programId, _shaderGeometry);
            glDetachShader(_programId, _shaderCompute);
        }

        private void DeleteShadersID()
        {
            glDeleteShader(_shaderVertex);
            glDeleteShader(_shaderFragment);
            glDeleteShader(_shaderGeometry);
            glDeleteShader(_shaderCompute);
        }
    }
}